source/common/http/conn_pool_grid.cc

```diff
@@ -0,0 +1,185 @@
+                                                     Upstream::HostDescriptionConstSharedPtr host,
+                                                     const StreamInfo::StreamInfo& info,
+                                                     absl::optional<Http::Protocol> protocol) {
+  // Right now, connections are tried serially, so any successful stream
+  // creation should be passed up to the original caller.
+  ConnectionPool::Callbacks& callbacks = inner_callbacks_;
+  delete this;
+  return callbacks.onPoolReady(encoder, host, info, protocol);
+}
+
+void ConnectivityGrid::WrapperCallbacks::cancel(Envoy::ConnectionPool::CancelPolicy cancel_policy) {
+  // If the newStream caller cancels the stream request, pass the cancellation on
+  // to the active pool.
+  cancellable_->cancel(cancel_policy);
+  delete this;
+}
+
+ConnectivityGrid::ConnectivityGrid(
+    Event::Dispatcher& dispatcher, Random::RandomGenerator& random_generator,
+    Upstream::HostConstSharedPtr host, Upstream::ResourcePriority priority,
+    const Network::ConnectionSocket::OptionsSharedPtr& options,
+    const Network::TransportSocketOptionsSharedPtr& transport_socket_options,
+    Upstream::ClusterConnectivityState& state, TimeSource& time_source,
+    ConnectivityOptions connectivity_options)
+    : dispatcher_(dispatcher), random_generator_(random_generator), host_(host),
+      priority_(priority), options_(options), transport_socket_options_(transport_socket_options),
+      state_(state), time_source_(time_source) {
+  // TODO(#15649) support v6/v4, WiFi/cellular.
+  ASSERT(connectivity_options.protocols_.size() == 3);
+  ASSERT(contains(connectivity_options.protocols_,
+                  {Http::Protocol::Http11, Http::Protocol::Http2, Http::Protocol::Http3}));
+}
+
+ConnectivityGrid::~ConnectivityGrid() {
+  // Ignore drained callbacks during teardown.
+  destroying_ = true;
+}
+
+absl::optional<ConnectivityGrid::PoolIterator> ConnectivityGrid::createNextPool() {
+  // Pools are created by newStream, which should not be called during draining.
+  ASSERT(drained_callbacks_.empty());
+  // Right now, only H3 and ALPN are supported, so if there are 2 pools we're done.
+  if (pools_.size() == 2 || !drained_callbacks_.empty()) {
+    return absl::nullopt;
+  }
+
+  // HTTP/3 is hard-coded as higher priority, H2 as secondary.
+  if (pools_.empty()) {
+    pools_.push_back(Http3::allocateConnPool(dispatcher_, random_generator_, host_, priority_,
+                                             options_, transport_socket_options_, state_,
+                                             time_source_));
+    return pools_.begin();
+  }
+  pools_.push_back(std::make_unique<HttpConnPoolImplMixed>(dispatcher_, random_generator_, host_,
+                                                           priority_, options_,
+                                                           transport_socket_options_, state_));
+  return pools_.begin()++;
```

Comment:
++ modifies the local copy of the iterator, how about `return std::next(pools_.begin();`